### PR TITLE
SW-5880 Added support for returning historical summary data

### PR DESCRIPTION
**Pull request by tommylau523**

### CODE REVIEW

**Comment by tommylau523** [comment]
* **#2636** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2636?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2636?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt [hunk 1]
```diff
@@ -75,8 +83,35 @@ class ObservationResultsStore(private val dslContext: DSLContext) {
     return fetchByCondition(OBSERVATIONS.plantingSites.ORGANIZATION_ID.eq(organizationId), limit)
   }
 
-  fun fetchSummaryForPlantingSite(
+  /**
+   * Retrieves summary for a planting site by aggregating latest observation data per subzone.
+   * Returns the latest value if a max observation time is not provided. Otherwise, return the
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt]: Nit: "Otherwise, return" -> "Otherwise, returns"

File: src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt [hunk 2]
```diff
@@ -75,8 +83,35 @@ class ObservationResultsStore(private val dslContext: DSLContext) {
     return fetchByCondition(OBSERVATIONS.plantingSites.ORGANIZATION_ID.eq(organizationId), limit)
   }
 
-  fun fetchSummaryForPlantingSite(
+  /**
+   * Retrieves summary for a planting site by aggregating latest observation data per subzone.
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt]: This comment suggests it should return one summary, but it seems to return a list of them with, I guess, the earliest observations successively removed. I admit I'm a little fuzzy on what the intended result is here.

File: src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt [hunk 3]
```diff
@@ -264,14 +264,37 @@ class ObservationResultsStoreTest : DatabaseTest(), RunsAsUser {
...
+            importObservationsCsv(prefix, numSpecies, it, time) to time
           }
-      assertNull(summaryBeforeObservations, "No observations made yet.")
-      assertSummary(prefix, summaries)
+
+      val summaries = resultsStore.fetchSummariesForPlantingSite(plantingSiteId)
+      assertEquals(
+          emptyList<ObservationRollupResultsModel>(),
+          summaryBeforeObservations,
+          "No observations made yet.")
```
Comment (sgrimm) [inline src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt]: Maybe put this above the `val observationTimes` -- I found it slightly hard to follow since it came between `val summaries` and the first assertion on `summaries`.

File: src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt [hunk 4]
```diff
@@ -264,14 +264,37 @@ class ObservationResultsStoreTest : DatabaseTest(), RunsAsUser {
         sizeMeters: Int,
     ) {
       importSiteFromCsvFile(prefix, sizeMeters)
-      val summaryBeforeObservations = resultsStore.fetchSummaryForPlantingSite(plantingSiteId)
-      val summaries =
+      val summaryBeforeObservations = resultsStore.fetchSummariesForPlantingSite(plantingSiteId)
+
+      val observationTimes =
```
Comment (sgrimm) [inline src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt:273]: Maybe this is in preparation for a subsequent change, but it doesn't look like the observation IDs are ever used for anything here. Could this be a `List<Instant>` instead of a `List<Pair<ObservationId, Instant>>`?

File: src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt [hunk 5]
```diff
@@ -264,14 +264,37 @@ class ObservationResultsStoreTest : DatabaseTest(), RunsAsUser {
...
+      assertEquals(
+          emptyList<ObservationRollupResultsModel>(),
+          summaryBeforeObservations,
+          "No observations made yet.")
+      assertSummary(prefix, summaries.filterNotNull().reversed())
+
+      assertEquals(
+          summaries.take(2),
+          resultsStore.fetchSummariesForPlantingSite(plantingSiteId, limit = 2),
+          "Partial summaries via limit should contain the latest observations.")
```
Comment (sgrimm) [inline src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt:286]: Is this really asserting what it says it is? I think this is just testing that the summaries are returned in a stable order and that the `limit` argument gives you the ones at the beginning of the list, not that the order has anything to do with the observation times.

**Review by sgrimm — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt (hunk 2 above)
Comment (tommylau523) [reply src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt]: Yes but with most recent observation removed instead! This is intended to support querying summary data trend. I have updated the docs for this function!

File: src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt (hunk 5 above)
Comment (tommylau523) [reply src/test/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStoreTest.kt:286]: The beginning of the list should have the most recent summary. Since observations were fetched ordered descending by completion time. Assuming maxCompletionTime is not set, the first entry will use completed observations up to the latest,  second will use up to second latest, etc. until will hit the limit

File: src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt [hunk 6]
```diff
@@ -75,8 +83,38 @@ class ObservationResultsStore(private val dslContext: DSLContext) {
...
+        fetchByPlantingSiteId(plantingSiteId, maxCompletionTime = maxCompletionTime).filter {
+          it.completedTime != null
+        }
+
+    val numObservations = completedObservations.size
+    val depth = max(0, limit?.let { min(it, numObservations) } ?: numObservations)
+
+    // Remove observations one at a time, to build a historical summary
+    return List(depth) {
+          val observations = completedObservations.takeLast(numObservations - it)
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt:109]: `.drop(it)` would make the intent a little clearer than `.takeLast(numObservations - it)`.

**Review by sgrimm — APPROVED** [review]

**Comment by tommylau523** [comment]
### Merge activity

* **Nov 26, 3:53 PM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/2636).
